saleor/plugins/webhook/shipping.py

```diff
@@ -0,0 +1,161 @@
+import base64
+import logging
+from collections import defaultdict
+from typing import TYPE_CHECKING, Any, Callable, Dict, List
+
+import graphene
+from django.core.cache import cache
+from django.db.models import QuerySet
+from prices import Money
+
+from ...shipping.interface import ShippingMethodData
+from ..base_plugin import ExcludedShippingMethod
+from .const import CACHE_EXCLUDED_SHIPPING_TIME, EXCLUDED_SHIPPING_REQUEST_TIMEOUT
+from .tasks import _get_webhooks_for_event, trigger_webhook_sync
+from .utils import APP_ID_PREFIX
+
+if TYPE_CHECKING:
+    from ...app.models import App
+
+
+logger = logging.getLogger(__name__)
+
+
+def to_shipping_app_id(app: "App", shipping_method_id: str) -> "str":
+    return base64.b64encode(
+        str.encode(f"{APP_ID_PREFIX}:{app.pk}:{shipping_method_id}")
+    ).decode("utf-8")
+
+
+def parse_list_shipping_methods_response(
+    response_data: Any, app: "App"
+) -> List["ShippingMethodData"]:
+    shipping_methods = []
+    for shipping_method_data in response_data:
+        method_id = shipping_method_data.get("id")
+        method_name = shipping_method_data.get("name")
+        method_amount = shipping_method_data.get("amount")
+        method_currency = shipping_method_data.get("currency")
+        method_maximum_delivery_days = shipping_method_data.get("maximum_delivery_days")
+
+        shipping_methods.append(
+            ShippingMethodData(
+                id=to_shipping_app_id(app, method_id),
+                name=method_name,
+                price=Money(method_amount, method_currency),
+                maximum_delivery_days=method_maximum_delivery_days,
+            )
+        )
+    return shipping_methods
+
+
+def get_excluded_shipping_methods_or_fetch(
+    webhooks: QuerySet, event_type: str, payload: str, cache_key: str
+) -> Dict[str, List[ExcludedShippingMethod]]:
+    """Return data of all excluded shipping methods.
+
+    The data will be fetched from the cache. If missing it will fetch it from all
+    defined webhooks by calling a request to each of them one by one.
+    """
+    cached_data = cache.get(cache_key)
+    if cached_data:
+        cached_payload, excluded_shipping_methods = cached_data
+        if payload == cached_payload:
+            return parse_excluded_shipping_methods(excluded_shipping_methods)
+
+    excluded_methods = []
+    # Gather responses from webhooks
+    for webhook in webhooks:
+        response_data = trigger_webhook_sync(
+            event_type,
+            payload,
+            webhook.app,
+            EXCLUDED_SHIPPING_REQUEST_TIMEOUT,
+        )
+        # TODO handle a case when we didn't receive a proper response
```

Comment:
Do we have a task for this TODO?

Reply:
I don't think it's a TODO anymore, as the section that made you laugh is already covering pretty much all possible exceptions

Reply:
So I would remove this TODO